websocket/polygon_test.go

```diff
@@ -1,99 +1,57 @@
-		ParseData: true, // comment for raw data handling
-		Log:       log,
-	})
-	if err != nil {
-		log.Fatal(err)
-	}
-	defer c.Close()
-
-	c.Close() // this shouldn't panic
-	if err := c.Connect(); err != nil {
-		log.Fatal(err)
-	}
-
-	// calling connect again shouldn't panic or data race
-	if err := c.Connect(); err != nil {
-		log.Fatal(err)
-	}
-
-	go printOutput(c) // comment for raw data handling
-	// go printRawOutput(c) // uncomment for raw data handling
-
-	if err := c.Subscribe(polygonws.StocksSecAggs, "AAPL", "MSFT"); err != nil {
-		log.Error(err)
-	}
-
-	time.Sleep(5 * time.Second)
-	if err := c.Subscribe(polygonws.StocksTrades, "*"); err != nil {
-		log.Error(err)
-	}
-
-	time.Sleep(250 * time.Millisecond)
-	if err := c.Unsubscribe(polygonws.StocksTrades, "*"); err != nil {
-		log.Error(err)
-	}
-
-	time.Sleep(5 * time.Second)
-	if err := c.Unsubscribe(polygonws.StocksSecAggs, "MSFT"); err != nil {
-		log.Error(err)
-	}
-
-	for {
-		select {
-		case <-ctx.Done():
-			return
-		}
-	}
+func TestSupportsTopic(t *testing.T) {
+	assert.Equal(t, true, supportsTopic("stocks", Topic(12)))
+	assert.Equal(t, false, supportsTopic("stocks", Topic(40)))
+	assert.Equal(t, true, supportsTopic("options", Topic(31)))
+	assert.Equal(t, false, supportsTopic("options", Topic(40)))
+	assert.Equal(t, true, supportsTopic("forex", Topic(52)))
+	assert.Equal(t, false, supportsTopic("forex", Topic(40)))
+	assert.Equal(t, true, supportsTopic("crypto", Topic(74)))
+	assert.Equal(t, false, supportsTopic("crypto", Topic(40)))
+	assert.Equal(t, false, supportsTopic("fakeMarket", Topic(15)))
 }
 
-func printOutput(c *polygonws.Client) {
-	for {
-		out := c.Output()
-		if out == nil {
-			break
-		}
-		fmt.Println(out)
-	}
+func TestGetParams(t *testing.T) {
+	p1, _ := getParams("stocks", Topic(12), "AAPL", "GME", "HOOD")
+	p2, _ := getParams("stocks", Topic(12))
+	s1 := "AM.AAPL,AM.GME,AM.HOOD"
+	s2 := "AM.*"
+	assert.Equal(t, p1, s1)
+	assert.Equal(t, p2, s2)
 }
+func TestSubscriptions(t *testing.T) {
+	c, _ := New(Config{
+		APIKey:    "test",
+		Feed:      RealTime,
+		Market:    Stocks,
+		ParseData: true,
+		Log:       logrus.New(),
+	})
 
-//nolint:deadcode
-func printRawOutput(c *polygonws.Client) {
-	for {
-		out := c.Output()
-		if out == nil {
-			break
-		}
-		if b, ok := out.(json.RawMessage); ok {
-			fmt.Println(string(b))
-		}
-	}
+	c.setSubscriptions(Topic(12), "AAPL", "TSLA")
+	_, aapl := c.subscriptions["AM"]["AAPL"]
+	_, tsla := c.subscriptions["AM"]["TSLA"]
+	assert.Equal(t, true, aapl)
+	assert.Equal(t, true, tsla)
+	c.deleteSubscriptions(Topic(12), "AAPL", "NFLX")
+	_, aapl = c.subscriptions["AM"]["AAPL"]
+	assert.Equal(t, false, aapl)
+	c.setSubscriptions(Topic(12), "*")
+	_, all := c.subscriptions["AM"]["*"]
+	_, tsla = c.subscriptions["AM"]["TSLA"]
+	assert.Equal(t, false, tsla)
+	assert.Equal(t, true, all)
+	_, trade := c.subscriptions["T"]
+	assert.Equal(t, false, trade)
+	c.deleteSubscriptions(Topic(13), "RDFN")
```

Comment:
Is there anything we can assert if there is a delete here? I get that `RDFN` did not exist. 

Maybe it means we need to log that it wasn't in the subscriptions set, and for the test we could possibly get the length of the set before the delete and then assert that the length does not change after the delete

Reply:
Good point. I'll add a test for that.

Reply:
Actually in this case, the set does not exist before line 54. This is testing the edge case where a user tries to delete from a topic that hasn't been instantiated yet. This is why I run assert.Equal(t, true, trade) afterwards to make sure that now c.subscriptions["T"] has been created.

Reply:
oh that makes sense, nice